Can't implement Drop on type with const_evaluatable_checked where bound

This issue seems to be specific to the `Drop` trait.

```rust
#![feature(const_generics)]
#![feature(const_evaluatable_checked)]
#![allow(incomplete_features)]

struct Bar<const N: usize>
    where [(); N - 1]:
{
    x: [u8; N - 1]
}

impl<const N: usize> Drop for Bar<N>
    where [(); N - 1]:
{
    fn drop(&mut self) {
    }
}
```

```
error[E0367]: `Drop` impl requires `[(); _]: '<empty>` but the struct it is implemented for does not
  --> src/lib.rs:12:11
   |
12 |     where [(); N - 1]:
   |           ^^^^^^^^^^^
   |
note: the implementor must specify the same requirement
  --> src/lib.rs:5:1
   |
5  | / struct Bar<const N: usize>
6  | |     where [(); N - 1]:
7  | | {
8  | |     x: [u8; N - 1]
9  | | }
   | |_^

error[E0367]: `Drop` impl requires `the constant `<Bar<N> as Drop>::{constant#0}` can be evaluated` but the struct it is implemented for does not
  --> src/lib.rs:12:16
   |
12 |     where [(); N - 1]:
   |                ^^^^^
   |
note: the implementor must specify the same requirement
  --> src/lib.rs:5:1
   |
5  | / struct Bar<const N: usize>
6  | |     where [(); N - 1]:
7  | | {
8  | |     x: [u8; N - 1]
9  | | }
   | |_^

```

`rustc --version --verbose`:
```
rustc 1.50.0-nightly (fe982319a 2020-11-19)
binary: rustc
commit-hash: fe982319aa0aa5bbfc2795791a753832292bd2ba
commit-date: 2020-11-19
host: x86_64-unknown-linux-gnu
release: 1.50.0-nightly
```

@rustbot label A-const-generics F-const_evaluatable_checked T-compiler

--- a/compiler/rustc_typeck/src/check/dropck.rs
+++ b/compiler/rustc_typeck/src/check/dropck.rs
@@ -217,10 +217,10 @@ fn ensure_drop_predicates_are_implied_by_item_defn<'tcx>(
         // repeated `.iter().any(..)` calls.
 
         // This closure is a more robust way to check `Predicate` equality
-        // than simple `==` checks (which were the previous implementation).
-        // It relies on `ty::relate` for `TraitPredicate` and `ProjectionPredicate`
-        // (which implement the Relate trait), while delegating on simple equality
-        // for the other `Predicate`.
+        // than simple `==` checks (which were the previous implementation). It relies on
+        // `ty::relate` for `TraitPredicate`, `ProjectionPredicate`, `ConstEvaluatable`
+        // `TypeOutlives` and `TypeWellFormedFromEnv` (which implement the Relate trait),
+        // while delegating on simple equality for the other `Predicate`.
         // This implementation solves (Issue #59497) and (Issue #58311).
         // It is unclear to me at the moment whether the approach based on `relate`
         // could be extended easily also to the other `Predicate`.
@@ -235,6 +235,17 @@ fn ensure_drop_predicates_are_implied_by_item_defn<'tcx>(
                 (ty::PredicateKind::Projection(a), ty::PredicateKind::Projection(b)) => {
                     relator.relate(predicate.rebind(a), p.rebind(b)).is_ok()
                 }
+                (
+                    ty::PredicateKind::ConstEvaluatable(def_a, substs_a),
+                    ty::PredicateKind::ConstEvaluatable(def_b, substs_b),
+                ) => tcx.try_unify_abstract_consts(((def_a, substs_a), (def_b, substs_b))),
+                (ty::PredicateKind::TypeOutlives(a), ty::PredicateKind::TypeOutlives(b)) => {
+                    relator.relate(predicate.rebind(a.0), p.rebind(b.0)).is_ok()
+                }
+                (
+                    ty::PredicateKind::TypeWellFormedFromEnv(a),
+                    ty::PredicateKind::TypeWellFormedFromEnv(b),
+                ) => relator.relate(predicate.rebind(a), p.rebind(b)).is_ok(),
                 _ => predicate == p,
             }
         };

--- a/compiler/rustc_typeck/src/check/dropck.rs
+++ b/compiler/rustc_typeck/src/check/dropck.rs
@@ -217,9 +217,9 @@ fn ensure_drop_predicates_are_implied_by_item_defn<'tcx>(
         // repeated `.iter().any(..)` calls.
 
         // This closure is a more robust way to check `Predicate` equality
-        // than simple `==` checks (which were the previous implementation). It relies on
-        // `ty::relate` for `TraitPredicate`, `ProjectionPredicate`, `ConstEvaluatable`
-        // `TypeOutlives` and `TypeWellFormedFromEnv` (which implement the Relate trait),
+        // than simple `==` checks (which were the previous implementation).
+        // It relies on `ty::relate` for `TraitPredicate`, `ProjectionPredicate`,
+        // `ConstEvaluatable` and `TypeOutlives` (which implement the Relate trait),
         // while delegating on simple equality for the other `Predicate`.
         // This implementation solves (Issue #59497) and (Issue #58311).
         // It is unclear to me at the moment whether the approach based on `relate`
@@ -242,10 +242,6 @@ fn ensure_drop_predicates_are_implied_by_item_defn<'tcx>(
                 (ty::PredicateKind::TypeOutlives(a), ty::PredicateKind::TypeOutlives(b)) => {
                     relator.relate(predicate.rebind(a.0), p.rebind(b.0)).is_ok()
                 }
-                (
-                    ty::PredicateKind::TypeWellFormedFromEnv(a),
-                    ty::PredicateKind::TypeWellFormedFromEnv(b),
-                ) => relator.relate(predicate.rebind(a), p.rebind(b)).is_ok(),
                 _ => predicate == p,
             }
         };

--- a/src/test/ui/const-generics/const_evaluatable_checked/drop_impl.rs
+++ b/src/test/ui/const-generics/const_evaluatable_checked/drop_impl.rs
@@ -1,0 +1,16 @@
+//check-pass
+#![feature(const_generics, const_evaluatable_checked)]
+#![allow(incomplete_features)]
+
+struct Foo<const N: usize>
+where
+    [(); N + 1]: ;
+
+impl<const N: usize> Drop for Foo<N>
+where
+    [(); N + 1]: ,
+{
+    fn drop(&mut self) {}
+}
+
+fn main() {}

